File: CPP/003_Program/main.cpp
#include <iostream>
#include <string>
using namespace std;

/**
 * Class Example - OOP Demonstration
 * Shows classes, objects, constructors, and methods
 */

class Student {
private:
    string name;
    int age;
    double gpa;
    int studentId;

public:
    // Constructor
    Student(string n, int a, double g, int id) {
        name = n;
        age = a;
        gpa = g;
        studentId = id;
    }

    // Default constructor
    Student() {
        name = "Unknown";
        age = 0;
        gpa = 0.0;
        studentId = 0;
    }

    // Getter methods
    string getName() { return name; }
    int getAge() { return age; }
    double getGPA() { return gpa; }
    int getStudentId() { return studentId; }

    // Setter methods
    void setName(string n) { name = n; }
    void setAge(int a) { age = a; }
    void setGPA(double g) { gpa = g; }
    void setStudentId(int id) { studentId = id; }

    // Method to display student information
    void display() {
        cout << "\n--- Student Information ---" << endl;
        cout << "ID: " << studentId << endl;
        cout << "Name: " << name << endl;
        cout << "Age: " << age << endl;
        cout << "GPA: " << gpa << endl;
        cout << "--------------------------" << endl;
    }

    // Method to check if student has honors
    bool hasHonors() {
        return gpa >= 3.5;
    }

    // Method to update GPA
    void updateGPA(double newGPA) {
        if (newGPA >= 0.0 && newGPA <= 4.0) {
            gpa = newGPA;
            cout << "GPA updated successfully!" << endl;
        } else {
            cout << "Invalid GPA value!" << endl;
        }
    }
};

int main() {
    cout << "=== C++ Class Example ===" << endl;

    // Create student objects
    Student student1("Alice Johnson", 20, 3.8, 1001);
    Student student2("Bob Smith", 21, 3.2, 1002);
    Student student3;  // Using default constructor

    // Display student information
    student1.display();
    student2.display();

    // Check honors
    if (student1.hasHonors()) {
        cout << student1.getName() << " has honors!" << endl;
    }

    if (student2.hasHonors()) {
        cout << student2.getName() << " has honors!" << endl;
    } else {
        cout << student2.getName() << " does not have honors." << endl;
    }

    // Update student3 using setters
    cout << "\nUpdating student3 information..." << endl;
    student3.setName("Charlie Brown");
    student3.setAge(19);
    student3.setStudentId(1003);
    student3.updateGPA(3.6);

    student3.display();

    if (student3.hasHonors()) {
        cout << student3.getName() << " has honors!" << endl;
    }

    return 0;
}

File: Ruby/003_Program/main.rb
#!/usr/bin/env ruby

# Class Example - OOP Demonstration
# Shows classes, objects, attributes, and methods

class Student
  attr_accessor :name, :age, :gpa, :student_id

  def initialize(name, age, gpa, student_id)
    @name = name
    @age = age
    @gpa = gpa
    @student_id = student_id
  end

  def display
    puts "\n--- Student Information ---"
    puts "ID: #{@student_id}"
    puts "Name: #{@name}"
    puts "Age: #{@age}"
    puts "GPA: #{@gpa}"
    puts "--------------------------"
  end

  def honors?
    @gpa >= 3.5
  end

  def update_gpa(new_gpa)
    if (0.0..4.0).include?(new_gpa)
      @gpa = new_gpa
      puts "GPA updated successfully!"
    else
      puts "Invalid GPA value!"
    end
  end

  def to_s
    "Student: #{@name} (ID: #{@student_id})"
  end
end

# Inheritance example
class GraduateStudent < Student
  attr_accessor :thesis_topic

  def initialize(name, age, gpa, student_id, thesis_topic)
    super(name, age, gpa, student_id)
    @thesis_topic = thesis_topic
  end

  def display
    super
    puts "Thesis Topic: #{@thesis_topic}"
    puts "--------------------------"
  end
end

# Main program
puts "=== Ruby Class Example ==="

# Create student objects
student1 = Student.new("Alice Johnson", 20, 3.8, 1001)
student2 = Student.new("Bob Smith", 21, 3.2, 1002)
grad_student = GraduateStudent.new("Charlie Brown", 24, 3.9, 1003, "Machine Learning Applications")

# Display students
student1.display
student2.display
grad_student.display

# Check honors
[student1, student2, grad_student].each do |student|
  status = student.honors? ? "has honors" : "does not have honors"
  puts "#{student.name} #{status}"
end

# Update GPA
puts "\nUpdating #{student2.name}'s GPA..."
student2.update_gpa(3.7)
student2.display

if student2.honors?
  puts "#{student2.name} now has honors!"
end
